Consistent line height, other style fixes, better style guide

jrubenoff opened the pull request:
Closes #249, #241, #223, #204, and dobtco/screendoor-v2#3302

- Uses a global unitless `line-height`, and makes appropriate alignment adjustments

----

Also:

- `<dt>` is now bold for added clarity:

![screen shot 2016-04-01 at 1 30 24 pm](https://cloud.githubusercontent.com/assets/1328849/14219211/eaddcdea-f80d-11e5-89a8-74fb79a8d06a.png)

- Improved "dos and don'ts" pattern, and nicer-looking screenshots, in the style guide:

![screen shot 2016-04-01 at 1 31 03 pm](https://cloud.githubusercontent.com/assets/1328849/14219220/00da92a4-f80e-11e5-9bc5-bd319a1e315b.png)

- Tighter spacing for formatted lists:

![screen shot 2016-04-01 at 1 32 16 pm](https://cloud.githubusercontent.com/assets/1328849/14219255/2a9abab0-f80e-11e5-9213-02838d718d9c.png)

- New documentation for text contrast, small caps, links to external sites in dropdowns

ajb commented:
Can you talk more about the decision to add units to the global line height? I'm trying to square it with https://css-tricks.com/almanac/properties/l/line-height/#article-header-id-0.

Maybe we should discuss on Monday.

jrubenoff commented:
We use a set, predictable range of font sizes that are hardcoded into variables, so that rule doesn't really apply. If we had a bunch of exceptions in our codebase where we were styling elements in arbitrary font sizes and prioritized flexibility over predictability, that advice would make sense. The approach in this PR gives us greater control over the design.

We were already using set line heights for `h1`, `h2`, `h3`, etc., so this just extends that pattern to the whole codebase.

jrubenoff commented:
On call:

- [ ] Remove units from all line heights

jrubenoff commented:
Alright, removed the units from all the line-heights. `1.5` is a little too tight for our smaller sizes, so I adjusted the `$lineHeightSmall-` and `$unitlessLineHeightSmall-` variables accordingly. Also made a few more minor improvements to the style guide.

ajb commented on vendor/assets/stylesheets/dvl/components/page_header.scss:
```diff
@@ -99,13 +99,12 @@
   position: relative;
-  top: $rhythm * 0.5;
+  top: $lineHeightSmall / 2;
   left: $fontOpticalMarginH2 * -1;
```
I think we can change this to `left: 0`, based on what it looks like.

ajb commented on vendor/assets/stylesheets/dvl/components/navbar.scss:
```diff
@@ -110,9 +110,8 @@
-  line-height: $rhythm * 2;
-  font-size: $fontSmallest;
+  height: $lineHeightSmallest;
+  @include fontSmallestBody;
```
Here's how this badge looks in `master`:

<img width="249" alt="screen shot 2016-05-05 at 7 42 19 am" src="https://cloud.githubusercontent.com/assets/1270317/15046514/f0787e5c-1294-11e6-89c3-09b4cbad1150.png">

Here's how it looks here:

<img width="181" alt="screen shot 2016-05-05 at 7 42 40 am" src="https://cloud.githubusercontent.com/assets/1270317/15046516/f9989602-1294-11e6-9739-3c01b4feb064.png">

:'(

ajb commented on vendor/assets/stylesheets/dvl/components/sidebar_data.scss:
```diff
@@ -44,15 +45,22 @@
+  position: absolute;
+  top: 50%;
+  left: 0;
+  margin-top: $rhythm * -3;
```
Don't we have a mixin for this?

ajb replied:
The font size is `2.5rem`, so I believe the `margin-top` needs to be `-1.25rem`, not `-1.5rem` as it is.

ajb replied:
@jrubenoff: did you see this?

jrubenoff replied:
No, thanks for bringing it to my attention.

ajb commented on vendor/assets/stylesheets/dvl/components/sidebar_data.scss:
```diff
@@ -44,15 +45,22 @@
     width: $rhythm * 5;
+    left: $rhythm * 2;
+    .sidebar_data li:last-of-type & {
+      margin-top: $rhythm * -1.5;
```
I can't figure this out by looking at it. Explain?

ajb replied:
@jrubenoff: would appreciate if you can explain

jrubenoff replied:
The last sidebar item has no bottom padding, and the first sidebar item has no top padding. So, we need to adjust the position of the icon to vertically align it properly for each item.

jrubenoff replied:
I'll add a comment explaining this.

ajb commented on vendor/assets/stylesheets/dvl/core/forms.scss:
```diff
@@ -463,17 +463,19 @@ form {
   background-color: $errorColor;
   color: $white;
-  padding: $rhythm $rhythm * 2;
+  padding: ($rhythm / 2) ($rhythm * 1.5);
```
Much better :D

ajb commented:
Here are my comments as I went through the changed files and previewed the site. I also added some line comments where appropriate.

I think a good next step would be to talk in-person (or video) before making any more changes.

---

Contrary to https://github.com/dobtco/screendoor-v2/pull/3523#issuecomment-216679696, adding lines of code _does_ seem to be a trend: http://take.ms/hAGhF. This is my primary resistance to this PR.

Similarly, I'm not sure the 1.64 line height is worth the added complexity, the likelihood that it won't be applied consistently, and the added file size.

And while we're on the subject: these large, multi-issue PRs are really a nightmare to deal with. We've talked about this before; how can we avoid them in the future?

---

I found what I think is a page header alignment bug: http://take.ms/ZZYVS

I might have missed it... why are "dos" and "don'ts" screenshots, instead of live HTML? The screenshots look kinda crappy on my display... I think because they're non-retina?

This footer line-height bug still exists: http://take.ms/ctfCv

The sidebar has some behavior that we've tried to avoid in the past: http://take.ms/ktK3X. (This might be why: http://take.ms/USHFY)

The redesigned user block looks nice, but feels a little cramped. How about adding some more padding on the right side? http://take.ms/lripS

---

jrubenoff commented:
I don't share your concerns around the `page_header` and `sidebar_data`, but I made a few aesthetic changes to those components. I also reverted the line-height mixins.

ajb commented:
@jrubenoff: there are some merge conflicts that need to be fixed, now that I merged the other footer line height PR and the list styles PR.

Can you do that, and then maybe write up a new summary for this PR? I'm having trouble actually fguring out what this changes.

jrubenoff commented:
Done and done

ajb commented on vendor/assets/stylesheets/dvl/components/tooltips.scss:
```diff
@@ -10,7 +10,6 @@
   min-width: 40px;
   transition: transform 0.25s $easeOutBack, opacity 120ms ease-out;
   // Override styles on parent links
-  line-height: $lineHeight;
```
I'm pretty sure this was important to have in here, as per the comment on line 12.

jrubenoff replied:
Thanks, fixed.

ajb commented:
Here's my list of what components changed:

- dropdown select
- dropdowns
- flashes
- navbar badge
- page header
- page header context
- page header tertiary
- sidebar data
- sidebar nav
- sidebar sub actions
- tooltips
- user block
- datetime picker
- `<dt>` smaller


I'm gonna preview all of these and look at the changes. I also left some new inline comments.

ajb commented:
Here's all I have for now:

- ![img](http://take.ms/9RvlC1)
- [ ] Should the `<dt>` really be `$fontSmallest`? I thought we should use it sparingly.

Thanks for sticking with this... 😅

jrubenoff commented:
Manually set the line-height for `navbar_badge`

`<dt>`s are `$fontSmaller`... maybe you misread?

ajb commented:
> <dt>s are $fontSmaller... maybe you misread?

Yeah, I did. But did they get smaller in this PR?

jrubenoff commented:
Yes, they were `fontSmall` before. Harder to distinguish.
